packages/protocol/contracts/verifiers/compose/ComposeVerifier.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import "../../common/EssentialContract.sol";
+import "../../common/LibStrings.sol";
+import "../IVerifier.sol";
+
+/// @title ComposeVerifier
+/// @notice This contract is an abstract verifier that composes multiple sub-verifiers to validate
+/// proofs.
+/// It ensures that a set of sub-proofs are verified by their respective verifiers before
+/// considering the overall proof as valid.
+/// @custom:security-contact security@taiko.xyz
+abstract contract ComposeVerifier is EssentialContract, IVerifier {
+    uint256[50] private __gap;
+
+    struct SubProof {
+        address verifier;
+        bytes proof;
+    }
+
+    event InvalidSubProof(address indexed verifier, bytes returnData);
+
+    error INVALID_CALLER();
+    error INVALID_VERIFIER();
+    error INSUFFICIENT_PROOF();
+
+    /// @notice Initializes the contract.
+    /// @param _owner The owner of this contract. msg.sender will be used if this value is zero.
+    /// @param _rollupAddressManager The address of the {AddressManager} contract.
+    function init(address _owner, address _rollupAddressManager) external initializer {
+        __Essential_init(_owner, _rollupAddressManager);
+    }
+
+    /// @notice Verifies one or more sub-proofs.
+    /// @param _ctx The context of the proof verification.
+    /// @param _tran The transition to verify.
+    /// @param _proof The proof to verify.
+    function verifyProof(
+        Context calldata _ctx,
+        TaikoData.Transition calldata _tran,
+        TaikoData.TierProof calldata _proof
+    )
+        external
+    {
+        if (!isCallerAuthorized(msg.sender)) revert INVALID_CALLER();
+
+        (address[] memory verifiers, uint256 threshold) = getSubVerifiersAndThreshold();
+        SubProof[] memory subproofs = abi.decode(_proof.data, (SubProof[]));
+        uint256 numVerified;
+
+        for (uint256 i; i < subproofs.length; ++i) {
+            if (subproofs[i].verifier == address(0)) revert INVALID_VERIFIER();
+
+            // find the verifier
+            bool verifierFound;
+            for (uint256 j; j < verifiers.length; ++j) {
+                if (verifiers[j] == subproofs[i].verifier) {
+                    verifierFound = true;
+                    verifiers[j] = address(0);
+                }
+            }
+
+            if (!verifierFound) revert INVALID_VERIFIER();
+
+            (bool success, bytes memory returnData) = subproofs[i].verifier.call(
+                abi.encodeCall(
+                    IVerifier.verifyProof,
```

Comment:
what was the reason again to not just call verifyProof directly on the address?

Reply:
Using the low level call may be an overkill here. I refactored the code to require the number of sub-proofs much match the threshold and each sub-proof must be verifiable.